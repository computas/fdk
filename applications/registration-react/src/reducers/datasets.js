import {
  DATASETS_REQUEST,
  DATASETS_SUCCESS,
  DATASETS_FAILURE
} from '../constants/ActionTypes';

export default function datasets(
  state = { isFetchingDatasets: false, datasetItems: null },
  action
) {
  switch (action.type) {
    case DATASETS_REQUEST:
      return {
        ...state,
        isFetchingDatasets: true
      };
    case DATASETS_SUCCESS:
      return {
        ...state,
        isFetchingDatasets: false,
        datasetItems: action.response.data
      };
    case DATASETS_FAILURE:
      return {
        ...state,
        isFetchingDatasets: false,
        datasetItems: null
      };
    default:
      return state;
  }
}
