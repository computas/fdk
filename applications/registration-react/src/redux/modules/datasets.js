import _ from 'lodash';
import {
  DATASETS_REQUEST,
  DATASETS_SUCCESS,
  DATASETS_FAILURE
} from '../../constants/ActionTypes';
import { fetchActions } from '../fetchActions';

function shouldFetch(metaState) {
  const threshold = 60 * 1000; // seconds
  return (
    !metaState ||
    (!metaState.isFetching &&
      (metaState.lastFetch || 0) < Date.now() - threshold)
  );
}

export function fetchDatasetsIfNeeded(catalogId) {
  return (dispatch, getState) =>
    shouldFetch(_.get(getState(), ['datasets', catalogId, 'meta'])) &&
    dispatch(
      fetchActions(`/catalogs/${catalogId}/datasets?size=100`, [
        { type: DATASETS_REQUEST, meta: { catalogId } },
        { type: DATASETS_SUCCESS, meta: { catalogId } },
        { type: DATASETS_FAILURE, meta: { catalogId } }
      ])
    );
}

const initialState = {};

export default function datasets(state = initialState, action) {
  switch (action.type) {
    case DATASETS_REQUEST:
      return {
        ...state,
        [action.meta.catalogId]: {
          meta: {
            isFetching: true,
            lastFetch: null
          }
        }
      };
    case DATASETS_SUCCESS:
      return {
        ...state,
        [action.meta.catalogId]: {
          items: _.get(action.payload, ['_embedded', 'datasets']),
          meta: {
            isFetching: false,
            lastFetch: Date.now()
          }
        }
      };
    case DATASETS_FAILURE:
      return {
        ...state,
        [action.meta.catalogId]: {
          meta: {
            isFetching: false,
            lastFetch: null
          }
        }
      };
    default:
      return state;
  }
}

export const getDatasetItemsByCatalogId = (datasets, catalogId) =>
  _.get(datasets, [catalogId, 'items']);
