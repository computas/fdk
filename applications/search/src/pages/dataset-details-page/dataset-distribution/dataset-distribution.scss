.fdk-container-detail-sample {
  border-left: 10px solid #000000;
}