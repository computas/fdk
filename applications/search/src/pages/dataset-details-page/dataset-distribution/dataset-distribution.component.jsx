import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import localization from '../../../lib/localization';
import { getDistributionTypeByUri } from '../../../redux/modules/distributionType';
import { getTranslateText } from '../../../lib/translateText';
import { DistributionFormat } from '../../../components/distribution-format/distribution-format.component';
import './dataset-distribution.scss';

export class DatasetDistribution extends React.Component {
  // eslint-disable-line react/prefer-stateless-function

  _renderType() {
    const { distributionTypeItems } = this.props;
    let { type } = this.props;
    if (type) {
      if (type !== 'API' && type !== 'Feed' && type !== 'Nedlastbar fil') {
        const distributionType = getDistributionTypeByUri(
          distributionTypeItems,
          type
        );
        if (distributionType !== null && distributionType.length > 0) {
          type = getTranslateText(distributionType[0].prefLabel);
        } else {
          type = null;
        }
      }
      return (
        <div>
          <h5 className="mt-5">{localization.dataset.distribution.type}</h5>
          <p className="fdk-ingress">{type}</p>
        </div>
      );
    }
    return null;
  }

  _renderFormats() {
    const { code, format } = this.props;
    const children = (items, code) =>
      items.map(item => {
        if (item !== null) {
          const formatArray = item.trim().split(',');
          return formatArray.map((item, index) => {
            if (item === null) {
              return null;
            }
            return (
              <DistributionFormat
                key={`dataset-distribution-format${index}`}
                code={code}
                text={item}
              />
            );
          });
        }
        return null;
      });

    if (format && format[0] !== null) {
      return (
        <div>
          <h5 className="mt-4">{localization.dataset.distribution.format}</h5>
          {children(format, code)}
        </div>
      );
    }
    return null;
  }

  _renderTilgangsURL() {
    const { accessUrl } = this.props;
    const children = items =>
      items.map((item, index) => (
        <a
          key={`dataset-distribution-accessurl-${index}`}
          className="dataset-distribution-accessurl"
          href={item}
        >
          {item}
          <i className="fa fa-external-link fdk-fa-right" />
        </a>
      ));

    if (accessUrl) {
      return (
        <div>
          <h5 className="mt-5">
            {localization.dataset.distribution.accessUrl}
          </h5>
          <p className="fdk-ingress">{children(accessUrl)}</p>
        </div>
      );
    }
    return null;
  }

  _renderLicense() {
    const { license } = this.props;
    if (license && license.uri) {
      return (
        <div>
          <h5 className="mt-5">{localization.dataset.distribution.license}</h5>
          <p className="fdk-ingress">
            {license &&
              license.uri &&
              license.prefLabel && (
                <a href={license.uri}>
                  {getTranslateText(license.prefLabel)}
                  <i className="fa fa-external-link fdk-fa-right" />
                </a>
              )}
            {license &&
              license.uri &&
              !license.prefLabel && (
                <a href={license.uri}>
                  {localization.dataset.distribution.licenseLinkDefault}
                  <i className="fa fa-external-link fdk-fa-right" />
                </a>
              )}
          </p>
        </div>
      );
    }
    return null;
  }

  _renderConformsTo() {
    const { conformsTo } = this.props;

    const children = items =>
      items.map(item => (
        <a key={item.uri} href={item.uri}>
          {item.prefLabel
            ? getTranslateText(item.prefLabel)
            : localization.dataset.distribution.standard}
          <i className="fa fa-external-link fdk-fa-right" />
        </a>
      ));

    if (conformsTo) {
      return (
        <div>
          <h5 className="mt-5">
            {localization.dataset.distribution.conformsTo}
          </h5>
          <p className="fdk-ingress">{children(conformsTo)}</p>
        </div>
      );
    }
    return null;
  }

  _renderDistributionPage() {
    const { page } = this.props;
    const children = items =>
      items.map(page => {
        if (page && page.uri) {
          return (
            <a key={page.uri} href={page.uri}>
              {page.prefLabel ? getTranslateText(page.prefLabel) : page.uri}
            </a>
          );
        }
        return null;
      });

    if (page) {
      return (
        <div>
          <h5 className="mt-5">{localization.dataset.distribution.page}</h5>
          <p className="fdk-ingress">{children(page)}</p>
        </div>
      );
    }
    return null;
  }

  render() {
    const { title, code } = this.props;
    const distributionClass = cx('fdk-container-detail', {
      'fdk-container-detail-unntatt-offentlig': code === 'NON_PUBLIC',
      'fdk-container-detail-begrenset': code === 'RESTRICTED',
      'fdk-container-detail-offentlig': code === 'PUBLIC',
      'fdk-container-detail-sample': code === 'SAMPLE'
    });
    return (
      <section className={distributionClass}>
        <h4>{title}</h4>
        {this.props.description && (
          <p className="fdk-ingress">{this.props.description}</p>
        )}
        {this._renderType()}
        {this._renderFormats()}
        {this._renderTilgangsURL()}
        {this._renderLicense()}
        {this._renderConformsTo()}
        {this._renderDistributionPage()}
      </section>
    );
  }
}

DatasetDistribution.defaultProps = {
  title: '',
  description: null,
  accessUrl: null,
  format: null,
  code: '',
  license: null,
  conformsTo: null,
  page: null,
  type: null,
  distributionTypeItems: null
};

DatasetDistribution.propTypes = {
  title: PropTypes.string,
  description: PropTypes.string,
  accessUrl: PropTypes.array,
  format: PropTypes.array,
  code: PropTypes.string,
  license: PropTypes.object,
  conformsTo: PropTypes.array,
  page: PropTypes.array,
  type: PropTypes.string,
  distributionTypeItems: PropTypes.array
};
