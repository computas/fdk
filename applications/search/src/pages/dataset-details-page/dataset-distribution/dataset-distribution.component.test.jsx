import React from 'react';
import { shallow } from 'enzyme';
import { DatasetDistribution } from './dataset-distribution.component';
import datasets from '../../../../test/fixtures/datasets';
// import openLicenses from '../../fixtures/openLicenses';

test('should render DatasetDistribution correctly with no props', () => {
  const minWrapper = shallow(<DatasetDistribution />);
  expect(minWrapper).toMatchSnapshot();
});

test('should render DatasetDistribution correctly', () => {
  const { _source } = datasets[0];
  const defaultProps = {
    title: _source.title.nb,
    description: _source.description.nb,
    accessUrl: _source.distribution[0].accessURL,
    format: _source.distribution[0].format,
    code: _source.distribution[0].code,
    license: _source.distribution[0].license,
    conformsTo: _source.conformsTo,
    page: _source.page
  };
  const wrapper = shallow(<DatasetDistribution {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render DatasetDistribution correctly', () => {
  const { _source } = datasets[1];
  const defaultProps = {
    title: _source.title.nb,
    description: _source.description.nb,
    accessUrl: _source.distribution[0].accessURL,
    format: _source.distribution[0].format,
    code: _source.distribution[0].code,
    license: _source.distribution[0].license,
    conformsTo: _source.conformsTo,
    page: _source.page,
    type: 'Feed'
  };
  const wrapper = shallow(<DatasetDistribution {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});
