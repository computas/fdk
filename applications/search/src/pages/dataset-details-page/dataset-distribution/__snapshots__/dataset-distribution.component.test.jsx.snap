// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DatasetDistribution correctly 1`] = `
<section
  className="fdk-container-detail"
>
  <h4>
    AAMarkagrensen Oslo Kommune og nærliggende kommuner
  </h4>
  <p
    className="fdk-ingress"
  >
    Datasettet avgrenser område for virkeområdet til lov 6. juni 2009 nr. 35 om naturområder i Oslo og nærliggende kommuner (markaloven) som trådte i kraft 1. september 2009. Markalovens virkeområde er fastsatt i forskrift 4. september 2015 nr. 1032 om justering av markagrensen fastlegger markalovens geografiske virkeområde med tilhørende kart.
  </p>
  <div>
    <h5
      className="mt-5"
    >
      AccessURL
    </h5>
    <p
      className="fdk-ingress"
    >
      <a
        className="dataset-distribution-accessurl"
        href="http://www.detteerenlenke.no/til-nedlasting"
        key="dataset-distribution-accessurl-0"
      >
        http://www.detteerenlenke.no/til-nedlasting
        <i
          className="fa fa-external-link fdk-fa-right"
        />
      </a>
      <a
        className="dataset-distribution-accessurl"
        href="http://www.detteerenannenlenke.no/til-en-annen-nedlasting
"
        key="dataset-distribution-accessurl-1"
      >
        http://www.detteerenannenlenke.no/til-en-annen-nedlasting

        <i
          className="fa fa-external-link fdk-fa-right"
        />
      </a>
      <a
        className="dataset-distribution-accessurl"
        href="http://www.detteerentredjelenke.no/til-en-tredje-nedlasting"
        key="dataset-distribution-accessurl-2"
      >
        http://www.detteerentredjelenke.no/til-en-tredje-nedlasting
        <i
          className="fa fa-external-link fdk-fa-right"
        />
      </a>
    </p>
  </div>
  <div>
    <h5
      className="mt-5"
    >
      License
    </h5>
    <p
      className="fdk-ingress"
    >
      <a
        href="https://www.kartverket.no/geodataarbeid/standarder/sosi/"
      >
        SOSI
        <i
          className="fa fa-external-link fdk-fa-right"
        />
      </a>
    </p>
  </div>
  <div>
    <h5
      className="mt-5"
    >
      Conforms to
    </h5>
    <p
      className="fdk-ingress"
    >
      <a
        href="https://www.kartverket.no/geodataarbeid/standarder/sosi/"
        key="https://www.kartverket.no/geodataarbeid/standarder/sosi/"
      >
        SOSI
        <i
          className="fa fa-external-link fdk-fa-right"
        />
      </a>
    </p>
  </div>
  <div>
    <h5
      className="mt-5"
    >
      More information about this distribution
    </h5>
    <p
      className="fdk-ingress"
    />
  </div>
</section>
`;

exports[`should render DatasetDistribution correctly 2`] = `
<section
  className="fdk-container-detail"
>
  <h4>
    AAMarkagrensen Oslo Kommune og nærliggende kommuner
  </h4>
  <p
    className="fdk-ingress"
  >
    Datasettet avgrenser område for virkeområdet til lov 6. juni 2009 nr. 35 om naturområder i Oslo og nærliggende kommuner (markaloven) som trådte i kraft 1. september 2009. Markalovens virkeområde er fastsatt i forskrift 4. september 2015 nr. 1032 om justering av markagrensen fastlegger markalovens geografiske virkeområde med tilhørende kart.
  </p>
  <div>
    <h5
      className="mt-5"
    >
      Type
    </h5>
    <p
      className="fdk-ingress"
    >
      Feed
    </p>
  </div>
  <div>
    <h5
      className="mt-4"
    >
      Format
    </h5>
    <DistributionFormat
      code=""
      key="dataset-distribution-format0"
      text="application/json"
      type={null}
    />
    <DistributionFormat
      code=""
      key="dataset-distribution-format0"
      text="xml"
      type={null}
    />
  </div>
  <div>
    <h5
      className="mt-5"
    >
      License
    </h5>
    <p
      className="fdk-ingress"
    >
      <a
        href="https://www.kartverket.no/geodataarbeid/standarder/sosi/"
      >
        SOSI
        <i
          className="fa fa-external-link fdk-fa-right"
        />
      </a>
    </p>
  </div>
  <div>
    <h5
      className="mt-5"
    >
      Conforms to
    </h5>
    <p
      className="fdk-ingress"
    >
      <a
        href="https://www.kartverket.no/geodataarbeid/standarder/sosi/"
        key="https://www.kartverket.no/geodataarbeid/standarder/sosi/"
      >
        SOSI
        <i
          className="fa fa-external-link fdk-fa-right"
        />
      </a>
    </p>
  </div>
  <div>
    <h5
      className="mt-5"
    >
      More information about this distribution
    </h5>
    <p
      className="fdk-ingress"
    />
  </div>
</section>
`;

exports[`should render DatasetDistribution correctly with no props 1`] = `
<section
  className="fdk-container-detail"
>
  <h4 />
</section>
`;
